src/vs/workbench/api/common/extHostTerminalService.ts

```diff
@@ -416,6 +418,28 @@ export abstract class BaseExtHostTerminalService extends Disposable implements I
 		this._proxy = extHostRpc.getProxy(MainContext.MainThreadTerminalService);
 		this._bufferer = new TerminalDataBufferer(this._proxy.$sendProcessData);
 		this._proxy.$registerProcessSupport(supportsProcesses);
+		this._extHostCommands.registerArgumentProcessor({
+			processArgument: arg => {
+				const deserialize = (arg: any) => {
+					const cast = arg as ISerializedTerminalInstanceContext;
+					return this._getTerminalById(cast.instanceId)?.value;
+				};
+				switch (arg?.$mid) {
+					case MarshalledId.TerminalContext: return deserialize(arg);
+					default: {
+						if (Array.isArray(arg)) {
+							return arg.map(e => {
+								switch (e?.$mid) {
+									case MarshalledId.TerminalContext: return deserialize(e);
+									default: return e;
+								}
+							});
```

Comment:
Perhaps do an in-place modification to avoid unnecessary garbage by cloning the array for arguments that aren't relevant for the terminal (since this is a very hot path)